### ISSUE TITLE
Remove non-HTML reporters from built mocha.js

### ISSUE THREAD

**Issue by ndhoule** [mochajs/mocha#1731]
See https://github.com/mochajs/mocha/pull/1727, https://github.com/mochajs/mocha/issues/509 for details.

tl;dr: Right now the built `mocha.js` file includes all of the reporters, even though they aren't useful in the browser.

Seems like we should expose a browser plugin/reporter interface before we remove these, seems like we'll break https://github.com/metaskills/mocha-phantomjs/ otherwise.

cc/ @metaskills @nathanboktae

**Comment by nathanboktae** [comment]
Yes removing the reporters will break mocha-phantomjs and [mocha-casperjs](https://github.com/nathanboktae/mocha-casperjs) - both use the compiled `mocha.js` and change the reporter as needed.

One thing to note too is mocha-phantomjs doesn't actually control the version of mocha. The user just points `mocha-phantomjs` at a flat file or server with their tests that has some version of mocha in it, so that makes life a bit tricker. `mocha-casperjs` does have control over it's version of mocha though. (Mocha runs in the phantomjs side in `mocha-casperjs`).

I've implemented ways to pull in 3rd party reporters for both projects (a bit different for both per differences in phantomjs vs the browser) - a bit hacky but it the [mocha-phantomjs mechanism](https://github.com/nathanboktae/mocha-phantomjs-core/blob/474a7604c9b396dcfbe60984defb0ebb7289bb6c/mocha-phantomjs-core.js#L133-L160) provides a way to require other reporters, namely the base reporter.

If per #509 the reporters were packaged separately, and this plugin mechanism was available in a robust, documented way where we could inject all the reporters included today and hook them up to the compiled version of mocha, that would work.

However the challenge is creating a a reporter that runs in node, the browser, and even phantomjs or another JavaScript environment. You can get there today with some "module magic dance" code, but perhaps we could help that situation out with a better plugin model / API.

Off the top of my head, it would look like this, for example:
```javascript
(function(factory) {
  if (typeof require === 'function' && typeof module === 'object' && typeof exports === 'object') {
    // CommonJS environmnent (node, browserify, phantomjs)
    module.exports = factory(require('./Base'), require('moment'))
  } else {
    // another environment with the compiled version of mocha (e.g. Browser)
    mocha.reporter(factory(Mocha.reporters.Base, moment))
  }
})(function(Base, moment) {
  var myReporter = function MyCustomReporter() { /* ... */ }
  // ....
  return myReporter
})
```
